File: code/Java/state/gumballmachine/WinnerState.java
package state.gumballmachine;

public class WinnerState implements State{
    private GumballMachine gumballMachine;

    public WinnerState(GumballMachine gm) {
        this.gumballMachine = gm;
    }

    @Override
    public void insertQuarter() {
        System.out.println("Please wait, we are already giving you a gumball");
    }

    @Override
    public void ejectQuarter() {
        System.out.println("Sorry, you already turned the crank");
    }

    @Override
    public void turnCrank() {
        System.out.println("Turning twice doesn't get you another gumball");
    }

    @Override
    public void dispense() {
        gumballMachine.releaseBall();
        if(gumballMachine.getCount() > 0) {
            gumballMachine.releaseBall();
            System.out.println("You are a winner! You got two gumballs for your quarter");
            if(gumballMachine.getCount() > 0)
                gumballMachine.setCurState(gumballMachine.getNoQuarterState());
            else {
                System.out.println("Oops, out of gumballs!");
                gumballMachine.setCurState(gumballMachine.getSoldOutState());
            }
        }
        else
            gumballMachine.setCurState(gumballMachine.getSoldOutState());
    }

    @Override
    public void refill() {    }

//    @Override
//    public String toString() {
//        return "WinnerState{" +
//                "gumballMachine=" + gumballMachine +
//                '}';
//    }
}

File: code/C++/state/winnerstate.cpp
#include "winnerstate.h"

WinnerState::WinnerState(GumballMachine * const gm) :gumballMachine(gm) {}

void WinnerState::insertQuarter()
{
	std::cout << "Please wait, we are already giving you a gumball" << std::endl;
}

void WinnerState::ejectQuarter()
{
	std::cout << "Sorry, you already turned the crank" << std::endl;
}

void WinnerState::turnCrank()
{
	std::cout << "Turning twice doesn't get you another gumball" << std::endl;
}

void WinnerState::dispense()
{
	this->gumballMachine->releaseBall();
	if (this->gumballMachine->getCount() > 0) {
		std::cout << "You are a winner! You got two gumballs for your quarter" << std::endl;
		this->gumballMachine->releaseBall();
		if (this->gumballMachine->getCount() > 0) {
			this->gumballMachine->setState(this->gumballMachine->getNoQuarterState());
		}
		else {
			std::cout << "Oops, out of gumballs!" << std::endl;
			this->gumballMachine->setState(this->gumballMachine->getSoldOutState());
		}
	}
	else {
		std::cout << "Oops, out of gumballs!" << std::endl;
		this->gumballMachine->setState(this->gumballMachine->getSoldOutState());
	}
}

void WinnerState::refill()
{

}

void WinnerState::display()
{
	std::cout << "WinnerState, count:" << this->gumballMachine->getCount() << std::endl;
}